src/Wasm/Exec/Eval.hs

```diff
@@ -479,6 +479,48 @@ step(Code cs cfg vs (e:es)) = (`runReaderT` cfg) $ do
                   Right () -> oldSize
           k (I32 result : vs') es
 
+        (MemoryFill, I32 0 : _ : I32 dst : vs') -> {-# SCC step_MemoryFill #-} do
+          inst    <- getFrameInst
+          mem     <- lift $ memory inst (0 @@ at)
+          -- Zero len with offset out-of-bounds at the end of memory is allowed
+          sz      <- lift $ lift $ Memory.size mem
+          if Memory.pageSize * sz >= dst
+            then k vs' es
+            else k vs' (Trapping (memoryErrorString Memory.MemoryBoundsError) @@ at : es)
+
+        (MemoryFill, I32 cnt : v : I32 dst : vs') -> {-# SCC step_MemoryFill #-} do
+          inst    <- getFrameInst
+          mem     <- lift $ memory inst (0 @@ at)
+          let [addr, count] = fromIntegral . i64_extend_u_i32 . fromIntegral <$> [dst, cnt]
+          if addr + count > 2^32
+            then k vs' (Trapping (memoryErrorString Memory.MemoryBoundsError) @@ at : es)
+            else do
+              eres <- lift $ lift $ runExceptT $ mapM_ (\off -> Memory.storePacked Pack8 mem addr off v) [0 .. pred cnt]
+              case eres of
+                Right () -> k vs' es
+                Left exn -> k vs' (Trapping (memoryErrorString exn) @@ at : es)
```

Comment:
This looks quite different from the reference code in
https://github.com/WebAssembly/spec/blob/ffb5e3b40db7a45237d4e2fc972c1552ac4abef5/interpreter/exec/eval.ml#L413-L428
(and likewise below)

The design of `winter` is to be a pretty much direct port, so that chances are good to get the semantics right, and also to be able to follow and apply subsequent changes easily. 

We have occasionally deviated, for performance reasons (e.g. https://www.joachim-breitner.de/blog/765-Faster_Winter_7__The_Zipper). Is this the reason here? Maybe worth adding a comment here… hmm, but I see you aren’t still implementing this to byte-wise operations here.

So why not just follow the logic of the reference code directly?

Reply:
Be cause I didn't even look at the other reference :-)

Will do.

Reply:
Looking at that reference implementation makes me think it is written with proof assistants in mind. It lowers the complex instructions to simple ones and stucturally "smaller" identical ones. There is a case of `MemoryCopy` that isn't even tail recursive. So to make this remotely performant we'll have to rewrite these anyway. Otherwise even the Rust-style "do prefix and suffix bitwise and the inner part u64-wise" will be executing faster.